### PR TITLE
Make GPT2Model a HybridBlock

### LINKED ISSUE #993: Export for GPT-2
## Description
I would like to be able to to export the GPT-2 model for text generation into MXNet Clojure. Would the same approach with the BERT export script work? https://github.com/dmlc/gluon-nlp/blob/master/scripts/bert/export.py

Looking for the best way to approach it - happy to help with this issue if I'm able.

### LINKED ISSUE #1015: prev_len in gpt.py
## Description
the gpt code in 
https://github.com/dmlc/gluon-nlp/blob/5e11334f5c00fd2875ab75c670b2595c560c30fc/scripts/text_generation/model/gpt.py#L258
seems to have a mistake.

I might be wrong, but I think the code `prev_len = states[0].shape[1]` means to return the length of previous key/value matrix (or the previous input tokens) . However it returns the number of multi-head (12 for the 124M gpt2 model).

### Error Message
NA

## To Reproduce
run scripts/text_generation/sequence_sampling.py script and print the output of `prev_len = states[0].shape[1]` at line 258 in scripts/text_generation/model/gpt.py.

## What have you tried to solve it?
1. change to `prev_len = states[0].shape[2]`

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
Fixes #993 
Fixes #1015

### Essentials ###
- [X] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

### Changes ###
- [X] Make GPT2Model a HybridBlock

## Comments ##
cc @dmlc/gluon-nlp-team @gigasquid

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1010/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1010/3/index.html

**Comment by gigasquid** [comment]
Thanks so much for helping with this @leezu.

I pulled the branch and tried to run the sequence sampling for gpt2 and got an error. I *think* it might actually be a problem on master with some refactoring (or I'm doing something wrong):

```10:10 $ python3  sequence_sampling.py  random-sample  --bos 'Deep learning and natural language processing'   --lm-model gpt2_345m 
Namespace(beam_size=5, bos='Deep learning and natural language processing', command='random-sample', gpu=0, lm_model='gpt2_345m', max_length=20, print_num=3, temperature=1.0, use_top_k=None)
Traceback (most recent call last):
  File "sequence_sampling.py", line 187, in <module>
    generate()
  File "sequence_sampling.py", line 146, in generate
    decoder, vocab = get_decoder_vocab(args.lm_model)
  File "sequence_sampling.py", line 116, in get_decoder_vocab
    ctx=ctx)
  File "/Users/cmeier/workspace/deep-learning/gluon-nlp/scripts/text_generation/model/__init__.py", line 64, in get_model
    return models[name](**kwargs)
  File "/Users/cmeier/workspace/deep-learning/gluon-nlp/scripts/text_generation/model/gpt.py", line 383, in gpt2_345m
    **kwargs)
  File "/Users/cmeier/workspace/deep-learning/gluon-nlp/scripts/text_generation/model/gpt.py", line 429, in _get_gpt2_model
    **kwargs)
  File "/Users/cmeier/workspace/deep-learning/gluon-nlp/scripts/text_generation/model/gpt.py", line 239, in __init__
    units=units, hidden_size=units * 4, prefix='ffn{}_'.format(i)))
  File "/Users/cmeier/workspace/deep-learning/gluon-nlp/scripts/text_generation/model/gpt.py", line 186, in __init__
    self._act = GELU(approximate=True)
  File "/usr/local/lib/python3.7/site-packages/gluonnlp/model/block.py", line 106, in __init__
    super(GELU, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'approximate'
```

The error does not occur if I go back to ` git checkout 4e555394baf557e5e55e1ae24a2147b03dce2213`

**Comment by leezu** [comment]
You need to use the development version of gluonnlp if you use the development version of the script. Use `pip install git+https://github.com/dmlc/gluon-nlp.git`

**Comment by mli** [comment]
Job PR-1010/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1010/4/index.html

File: scripts/text_generation/model/gpt.py [hunk 1]
```diff
@@ -253,15 +264,20 @@ def forward(self, data, states=None): # pylint: disable=arguments-differ
...
+                F.contrib.arange_like(data, axis=1).astype('int32'), prev_len)
         else:
-            prev_len = 0
-        assert seq_len + prev_len <= self._max_length
-        data_pos = mx.nd.arange(prev_len, prev_len + seq_len, ctx=data.context, dtype=np.float32)
-        data_pos = mx.nd.broadcast_axes(mx.nd.expand_dims(data_pos, axis=0),
-                                        axis=0, size=batch_size)
+            data_pos = F.contrib.arange_like(data, axis=1).astype('int32')
+        if F is mx.nd:
+            length = data.shape[1] + (states[0].shape[1] if states is not None else 0)
```
Comment (sxjscience) [inline scripts/text_generation/model/gpt.py]: We should use states[0].shape[2]

**Comment by mli** [comment]
Job PR-1010/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1010/8/index.html

**Pull request merged**